### ISSUE TITLE
[Windows + Dragon] Hang on talon start-up

### ISSUE THREAD

**Issue by mirober** [talonvoice/talon#195]
Log:
```
2021-01-03 10:44:56 DEBUG User modules loaded at 16.9530s
2021-01-03 10:44:57    IO Injecting draconity into App(pid=10080, "natspeak.exe"), runtime=0:00:59.826060
2021-01-03 10:44:57    IO Injected into App(pid=10080, "natspeak.exe") successfully
2021-01-03 10:45:56 DEBUG Gathering (7) metrics archives for upload
2021-01-03 10:45:57 DEBUG Metrics upload finished (lines=190 failed=False)
2021-01-03 11:09:47  INFO (SpeechSystem) Activating speech engine: DraconityEngine(dragon)
2021-01-03 11:09:49 WARNING [watchdog] "talon.windows.ui._on_event" @2.0s
   34:                     threading.py:912| 
   33:                     threading.py:954| 
   32:                     threading.py:892| 
   31:       talon\windows\winevents.py:166| 
   30: ------------------------------------# cron thread
   29:       talon\windows\winevents.py:141| 
   28:       talon\windows\winevents.py:216| 
   27: ------------------------------------# 'cron' main:call()
   26:      talon\scripting\dispatch.py:98 | 
   25:      talon\scripting\dispatch.py:133| 
   24:      talon\scripting\dispatch.py:124| 
   23:          talon\scripting\rctx.py:201| 
   22:              talon\windows\ui.py:80 | 
   21:      talon\scripting\dispatch.py:98 | 
   20:      talon\scripting\dispatch.py:133| 
   19:      talon\scripting\dispatch.py:124| 
   18:          talon\scripting\rctx.py:201| 
   17:         talon\scripting\scope.py:28 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:126| 
   13:          talon\scripting\rctx.py:201| 
   12:         talon\scripting\scope.py:129| 
   11:      talon\scripting\dispatch.py:98 | 
   10:      talon\scripting\dispatch.py:133| 
    9:      talon\scripting\dispatch.py:126| 
    8:          talon\scripting\rctx.py:201| 
    7:         talon\scripting\scope.py:53 | 
    6:      talon\scripting\registry.py:477| 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1: talon\scripting\speech_system.py:208| 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.scope.update" @2.0s
   34:                     threading.py:912| 
   33:                     threading.py:954| 
   32:                     threading.py:892| 
   31:       talon\windows\winevents.py:166| 
   30: ------------------------------------# cron thread
   29:       talon\windows\winevents.py:141| 
   28:       talon\windows\winevents.py:216| 
   27: ------------------------------------# 'cron' app:call()
   26:      talon\scripting\dispatch.py:98 | 
   25:      talon\scripting\dispatch.py:133| 
   24:      talon\scripting\dispatch.py:124| 
   23:          talon\scripting\rctx.py:201| 
   22:              talon\windows\ui.py:80 | 
   21:      talon\scripting\dispatch.py:98 | 
   20:      talon\scripting\dispatch.py:133| 
   19:      talon\scripting\dispatch.py:124| 
   18:          talon\scripting\rctx.py:201| 
   17:         talon\scripting\scope.py:28 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:126| 
   13:          talon\scripting\rctx.py:201| 
   12:         talon\scripting\scope.py:129| 
   11:      talon\scripting\dispatch.py:98 | 
   10:      talon\scripting\dispatch.py:133| 
    9:      talon\scripting\dispatch.py:126| 
    8:          talon\scripting\rctx.py:201| 
    7:         talon\scripting\scope.py:53 | 
    6:      talon\scripting\registry.py:477| 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1: talon\scripting\speech_system.py:208| 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.scope.update_one" @2.0s
   34:                     threading.py:912| 
   33:                     threading.py:954| 
   32:                     threading.py:892| 
   31:       talon\windows\winevents.py:166| 
   30: ------------------------------------# cron thread
   29:       talon\windows\winevents.py:141| 
   28:       talon\windows\winevents.py:216| 
   27: ------------------------------------# 'cron' main:call()
   26:      talon\scripting\dispatch.py:98 | 
   25:      talon\scripting\dispatch.py:133| 
   24:      talon\scripting\dispatch.py:124| 
   23:          talon\scripting\rctx.py:201| 
   22:              talon\windows\ui.py:80 | 
   21:      talon\scripting\dispatch.py:98 | 
   20:      talon\scripting\dispatch.py:133| 
   19:      talon\scripting\dispatch.py:124| 
   18:          talon\scripting\rctx.py:201| 
   17:         talon\scripting\scope.py:28 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:126| 
   13:          talon\scripting\rctx.py:201| 
   12:         talon\scripting\scope.py:129| 
   11:      talon\scripting\dispatch.py:98 | 
   10:      talon\scripting\dispatch.py:133| 
    9:      talon\scripting\dispatch.py:126| 
    8:          talon\scripting\rctx.py:201| 
    7:         talon\scripting\scope.py:53 | 
    6:      talon\scripting\registry.py:477| 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1: talon\scripting\speech_system.py:208| 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.scope._registry_update_contexts" @2.1s
   34:                     threading.py:912| 
   33:                     threading.py:954| 
   32:                     threading.py:892| 
   31:       talon\windows\winevents.py:166| 
   30: ------------------------------------# cron thread
   29:       talon\windows\winevents.py:141| 
   28:       talon\windows\winevents.py:216| 
   27: ------------------------------------# 'cron' main:call()
   26:      talon\scripting\dispatch.py:98 | 
   25:      talon\scripting\dispatch.py:133| 
   24:      talon\scripting\dispatch.py:124| 
   23:          talon\scripting\rctx.py:201| 
   22:              talon\windows\ui.py:80 | 
   21:      talon\scripting\dispatch.py:98 | 
   20:      talon\scripting\dispatch.py:133| 
   19:      talon\scripting\dispatch.py:124| 
   18:          talon\scripting\rctx.py:201| 
   17:         talon\scripting\scope.py:28 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:126| 
   13:          talon\scripting\rctx.py:201| 
   12:         talon\scripting\scope.py:129| 
   11:      talon\scripting\dispatch.py:98 | 
   10:      talon\scripting\dispatch.py:133| 
    9:      talon\scripting\dispatch.py:126| 
    8:          talon\scripting\rctx.py:201| 
    7:         talon\scripting\scope.py:53 | 
    6:      talon\scripting\registry.py:477| 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1: talon\scripting\speech_system.py:208| 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.dispatch.<lambda>" @2.1s
   25:                     threading.py:912| 
   24:                     threading.py:954| 
   23:                     threading.py:892| 
   22:                    talon\cron.py:110| 
   21: ------------------------------------# cron thread
   20:                    talon\cron.py:75 | 
   19:          talon\scripting\rctx.py:201| 
   18: ------------------------------------# 'cron' main:call()
   17:      talon\scripting\dispatch.py:92 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:124| 
   13:          talon\scripting\rctx.py:201| 
   12: talon\scripting\speech_system.py:312| 
   11: talon\scripting\speech_system.py:273| 
   10:      talon\scripting\dispatch.py:98 | 
    9:      talon\scripting\dispatch.py:133| 
    8:      talon\scripting\dispatch.py:126| 
    7:          talon\scripting\rctx.py:201| 
    6:         talon\scripting\scope.py:28 | 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1:         talon\scripting\scope.py:99 | 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.speech_system.engine_event" @2.1s
   25:                     threading.py:912| 
   24:                     threading.py:954| 
   23:                     threading.py:892| 
   22:                    talon\cron.py:110| 
   21: ------------------------------------# cron thread
   20:                    talon\cron.py:75 | 
   19:          talon\scripting\rctx.py:201| 
   18: ------------------------------------# 'cron' main:call()
   17:      talon\scripting\dispatch.py:92 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:124| 
   13:          talon\scripting\rctx.py:201| 
   12: talon\scripting\speech_system.py:312| 
   11: talon\scripting\speech_system.py:273| 
   10:      talon\scripting\dispatch.py:98 | 
    9:      talon\scripting\dispatch.py:133| 
    8:      talon\scripting\dispatch.py:126| 
    7:          talon\scripting\rctx.py:201| 
    6:         talon\scripting\scope.py:28 | 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1:         talon\scripting\scope.py:99 | 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.speech_system.update_grammars" @2.1s
   34:                     threading.py:912| 
   33:                     threading.py:954| 
   32:                     threading.py:892| 
   31:       talon\windows\winevents.py:166| 
   30: ------------------------------------# cron thread
   29:       talon\windows\winevents.py:141| 
   28:       talon\windows\winevents.py:216| 
   27: ------------------------------------# 'cron' main:call()
   26:      talon\scripting\dispatch.py:98 | 
   25:      talon\scripting\dispatch.py:133| 
   24:      talon\scripting\dispatch.py:124| 
   23:          talon\scripting\rctx.py:201| 
   22:              talon\windows\ui.py:80 | 
   21:      talon\scripting\dispatch.py:98 | 
   20:      talon\scripting\dispatch.py:133| 
   19:      talon\scripting\dispatch.py:124| 
   18:          talon\scripting\rctx.py:201| 
   17:         talon\scripting\scope.py:28 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:126| 
   13:          talon\scripting\rctx.py:201| 
   12:         talon\scripting\scope.py:129| 
   11:      talon\scripting\dispatch.py:98 | 
   10:      talon\scripting\dispatch.py:133| 
    9:      talon\scripting\dispatch.py:126| 
    8:          talon\scripting\rctx.py:201| 
    7:         talon\scripting\scope.py:53 | 
    6:      talon\scripting\registry.py:477| 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1: talon\scripting\speech_system.py:208| 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.scope.update" @2.0s
   25:                     threading.py:912| 
   24:                     threading.py:954| 
   23:                     threading.py:892| 
   22:                    talon\cron.py:110| 
   21: ------------------------------------# cron thread
   20:                    talon\cron.py:75 | 
   19:          talon\scripting\rctx.py:201| 
   18: ------------------------------------# 'cron' main:call()
   17:      talon\scripting\dispatch.py:92 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:124| 
   13:          talon\scripting\rctx.py:201| 
   12: talon\scripting\speech_system.py:312| 
   11: talon\scripting\speech_system.py:273| 
   10:      talon\scripting\dispatch.py:98 | 
    9:      talon\scripting\dispatch.py:133| 
    8:      talon\scripting\dispatch.py:126| 
    7:          talon\scripting\rctx.py:201| 
    6:         talon\scripting\scope.py:28 | 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1:         talon\scripting\scope.py:99 | 
2021-01-03 11:09:49 WARNING [watchdog] "talon.scripting.scope.update_one" @2.0s
   25:                     threading.py:912| 
   24:                     threading.py:954| 
   23:                     threading.py:892| 
   22:                    talon\cron.py:110| 
   21: ------------------------------------# cron thread
   20:                    talon\cron.py:75 | 
   19:          talon\scripting\rctx.py:201| 
   18: ------------------------------------# 'cron' main:call()
   17:      talon\scripting\dispatch.py:92 | 
   16:      talon\scripting\dispatch.py:98 | 
   15:      talon\scripting\dispatch.py:133| 
   14:      talon\scripting\dispatch.py:124| 
   13:          talon\scripting\rctx.py:201| 
   12: talon\scripting\speech_system.py:312| 
   11: talon\scripting\speech_system.py:273| 
   10:      talon\scripting\dispatch.py:98 | 
    9:      talon\scripting\dispatch.py:133| 
    8:      talon\scripting\dispatch.py:126| 
    7:          talon\scripting\rctx.py:201| 
    6:         talon\scripting\scope.py:28 | 
    5:      talon\scripting\dispatch.py:98 | 
    4:      talon\scripting\dispatch.py:133| 
    3:      talon\scripting\dispatch.py:126| 
    2:          talon\scripting\rctx.py:201| 
    1:         talon\scripting\scope.py:99 | 
2021-01-03 11:09:51 WARNING [watchdog] "talon.windows.ui._on_event" @4.0s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.scope.update" @4.0s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.scope.update_one" @4.1s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.scope._registry_update_contexts" @4.1s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.dispatch.<lambda>" @4.1s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.speech_system.engine_event" @4.1s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.speech_system.update_grammars" @4.1s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.scope.update" @4.0s (stalled)
2021-01-03 11:09:51 WARNING [watchdog] "talon.scripting.scope.update_one" @4.0s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.windows.ui._on_event" @6.0s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.scope.update" @6.0s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.scope.update_one" @6.0s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.scope._registry_update_contexts" @6.1s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.dispatch.<lambda>" @6.0s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.speech_system.engine_event" @6.1s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.speech_system.update_grammars" @6.1s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.scope.update" @6.0s (stalled)
2021-01-03 11:09:53 WARNING [watchdog] "talon.scripting.scope.update_one" @6.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.windows.ui._on_event" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.scope.update" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.scope.update_one" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.scope._registry_update_contexts" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.dispatch.<lambda>" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.speech_system.engine_event" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.speech_system.update_grammars" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.scope.update" @8.0s (stalled)
2021-01-03 11:09:55 WARNING [watchdog] "talon.scripting.scope.update_one" @8.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.windows.ui._on_event" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.scope.update" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.scope.update_one" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.scope._registry_update_contexts" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.dispatch.<lambda>" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.speech_system.engine_event" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.speech_system.update_grammars" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.scope.update" @10.0s (stalled)
2021-01-03 11:09:57 WARNING [watchdog] "talon.scripting.scope.update_one" @10.0s (stalled)
```
...

I was messing around unplugging and re-plugging my microphone at the time, but even once Dragon was back online and recognising commands talon was still stalled here. Had to kill the process and restart. 

Does this give enough information to debug? From the logs it looks like this has happened a couple of times in the last month or two so if I ran a build with extra logging it would probably catch it eventually.

**Comment by lunixbochs** [comment]
What's your talon version?

**Comment by mirober** [comment]
Latest beta - 0.1.3-19

**Comment by lunixbochs** [comment]
I just completely reworked speech_system locking to only lock tightly around critical points. This should be fixed in the next beta (and v0.1.4)
